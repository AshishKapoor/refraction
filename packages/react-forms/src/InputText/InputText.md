The `InputText` component is the base unit used to create most of the more
advanced input components.

It reproduces all the functionalities of a standard HTML `<input />` element,
with some useful additions.

```js
<form>
  <table className="DemoTable">
    <thead>
      <tr>
        <th />
        <th>Default</th>
        <th>Typed Text</th>
        <th>Disabled</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>No Label</td>
        <td>
          <InputText />
        </td>
        <td>
          <InputText defaultValue="Typed something" />
        </td>
        <td>
          <InputText disabled />
        </td>
      </tr>
      <tr>
        <td>With Placeholder</td>
        <td>
          <InputText placeholder="I'm a Label" />
        </td>
        <td>
          <InputText placeholder="I'm a Label" defaultValue="Typed something" />
        </td>
        <td>
          <InputText placeholder="I'm a Label" disabled />
        </td>
      </tr>
      <tr>
        <td>Large</td>
        <td>
          <InputText size="large" placeholder="I'm a Label" />
        </td>
        <td>
          <InputText
            size="large"
            placeholder="I'm a Label"
            defaultValue="Typed something"
          />
        </td>
        <td>
          <InputText size="large" placeholder="I'm a Label" disabled />
        </td>
      </tr>
      <tr>
        <td>Small</td>
        <td>
          <InputText size="small" placeholder="I'm a Label" />
        </td>
        <td>
          <InputText
            size="small"
            placeholder="I'm a Label"
            defaultValue="Typed something"
          />
        </td>
        <td>
          <InputText size="small" placeholder="I'm a Label" disabled />
        </td>
      </tr>
      <tr>
        <td>With addon</td>
        <td>
          <InputText
            placeholder="I'm a Label"
            renderAddon={({ onClick, marginRightClass }) => (
              <Icon
                className={marginRightClass}
                name="calendar"
                onClick={onClick}
              />
            )}
          />
        </td>
        <td>
          <InputText
            placeholder="I'm a Label"
            defaultValue="Typed something"
            renderAddon={({ onClick, marginRightClass }) => (
              <Icon
                className={marginRightClass}
                name="calendar"
                onClick={onClick}
              />
            )}
          />
        </td>
        <td>
          <InputText
            placeholder="I'm a Label"
            disabled
            renderAddon={({ onClick, marginRightClass }) => (
              <Icon
                className={marginRightClass}
                name="calendar"
                onClick={onClick}
              />
            )}
          />
        </td>
      </tr>
    </tbody>
  </table>
</form>
```

#### HTML form validation

The `InputText` component supports HTML form validation, with some improvements.  
The **invalid** state will not be visible until the `InputText`'s parent form is
submitted, this improves the user experience, since the user will only see validation
errors after they attempt to submit the form rather than while they are writing.

Additionally, it is possible to provide a `validationErrorMessage` property to
define a custom error message to display when the field is marked as invalid.

```js
<form onSubmit={evt => evt.preventDefault()}>
  <InputText
    required
    style={{ width: 400 }}
    placeholder="Press enter to trigger validation, write something to reset"
    validationErrorMessage="This is a custom error message"
  />
</form>
```

#### Addon

Is it possible to add an arbitrary element on the right of the `InputText` typing area
by providing a `renderAddon` property.

The property expects a function that takes as arguments an object containing an `onClick`
proprty. This property should be assigned to the addon element to make sure the
text fields get focused when its addon is clicked.

```js
<InputText
  renderAddon={({ onClick, marginRightClass }) => (
    <Icon className={marginRightClass} name="caret_down" onClick={onClick} />
  )}
/>
```
