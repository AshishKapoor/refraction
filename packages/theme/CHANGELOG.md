# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.5.1](https://github.com/quid/ui-framework/compare/v1.5.0...v1.5.1) (2018-12-06)


### Bug Fixes

* use resolutions to point to microbundle fork ([16819ab](https://github.com/quid/ui-framework/commit/16819ab))





# [1.5.0](https://github.com/quid/ui-framework/compare/v1.4.0...v1.5.0) (2018-12-06)


### Bug Fixes

* use testPathPattern rather than runTestsByPath ([2350a80](https://github.com/quid/ui-framework/commit/2350a80))





# [1.4.0](https://github.com/quid/ui-framework/compare/v1.3.1...v1.4.0) (2018-12-05)

**Note:** Version bump only for package @quid/theme





# [1.3.0](https://github.com/quid/ui-framework/compare/v1.2.1...v1.3.0) (2018-12-05)


### Bug Fixes

* properly render CSS ([05f2920](https://github.com/quid/ui-framework/commit/05f2920))
* textStyles is not part of the theme ([919b026](https://github.com/quid/ui-framework/commit/919b026))


### Features

* new theme and react-forms packages ([61a44a0](https://github.com/quid/ui-framework/commit/61a44a0))
