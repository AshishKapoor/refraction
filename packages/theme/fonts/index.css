@import './ibmplex/index.css';
@import './icons/index.css';
