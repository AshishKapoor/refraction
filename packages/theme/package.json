{
  "name": "@quid/theme",
  "version": "1.5.2",
  "description": "Colors, fonts and other styling related resources, consumable by Emotion",
  "main": "dist/index.js",
  "main:umd": "dist/index.umd.js",
  "module": "dist/index.es.js",
  "license": "MIT",
  "scripts": {
    "start": "microbundle watch",
    "prepare": "microbundle build && flow-copy-source --ignore '{__mocks__/*,*.test}.js' src dist",
    "test": "cd ../.. && yarn test --testPathPattern packages/theme"
  },
  "dependencies": {
    "@emotion/core": "^10.0.2",
    "@emotion/utils": "^0.11.1",
    "color": "^3.1.0",
    "emotion-theming": "^10.0.2",
    "react": "^16.3.0"
  },
  "devDependencies": {
    "flow-copy-source": "^2.0.2",
    "microbundle": "^0.8.3"
  }
}
