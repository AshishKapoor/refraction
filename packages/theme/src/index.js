// @flow
import React, { type Node } from 'react';
import { ThemeProvider } from 'emotion-theming';

import themeData from './theme';
import * as themes from './themes';
export { themes };
export { default as colors } from './colors';
export { default as sizes } from './sizes';
export { default as textStyles } from './textStyles';
export { default as withFallback } from './withFallback';

type Props = {
  theme?: 'dark' | 'light',
  children: Node,
};

const DEFAULT = 'light';
const QuidThemeProvider = ({ theme, children, ...props }: Props) => (
  <ThemeProvider
    {...props}
    theme={themeData[theme != null ? theme : DEFAULT]}
    children={children}
  />
);

export default QuidThemeProvider;
