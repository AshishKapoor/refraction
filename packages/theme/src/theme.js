/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import colors from './colors';
import * as themes from './themes';
import sizes from './sizes';

const theme = {
  light: {
    current: 'light',
    colors,
    sizes,
    ...themes.light,
  },
  dark: {
    current: 'dark',
    colors,
    sizes,
    ...themes.dark,
  },
};

export default theme;
