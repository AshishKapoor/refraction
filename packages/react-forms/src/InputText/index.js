// @flow
import * as React from 'react';
import styled from '@emotion/styled/macro';
import { textStyles, withFallback as wf } from '@quid/theme';
import InvalidHandler from '../InvalidHandler';

type RenderProp<P> = P => React.Node;

type Props = {
  type?: string,
  size?: 'regular' | 'large' | 'small',
  renderAddon?: RenderProp<{ onClick: (SyntheticEvent<Element>) => void }>,
  as?: string,
  onChange?: (SyntheticInputEvent<any>) => void,
  validationErrorMessage?: string,
};

const omit = obj => keys =>
  Object.keys(obj)
    .filter(key => !keys.includes(key))
    .reduce((acc, current) => {
      acc[current] = obj[current];
      return acc;
    }, {});

const include = obj => keys =>
  Object.keys(obj)
    .filter(key => keys.includes(key))
    .reduce((acc, current) => {
      acc[current] = obj[current];
      return acc;
    }, {});

// istanbul ignore next
const mergeRefs = (...refs: Array<any>) => (ref: any) => {
  refs.forEach(resolvableRef => {
    if (typeof resolvableRef === 'function') {
      resolvableRef(ref);
    } else if (resolvableRef != null) {
      (resolvableRef: any).current = ref;
    }
  });
};

// istanbul ignore next
const noop = () => undefined;

const INPUT_ATTRIBUTES = [
  'autoComplete',
  'autoFocus',
  'defaultValue',
  'form',
  'list',
  'min',
  'max',
  'name',
  'onChange',
  'placeholder',
  'readOnly',
  'required',
  'tabIndex',
  'type',
  'value',
];

const HEIGHT = {
  large: 50,
  small: 24,
  regular: 32,
};

const PADDING = {
  large: 15,
  small: 5,
  regular: 10,
};

const Input = styled.input`
  all: unset;
  min-width: 0;
  align-self: stretch;
  flex: 1;

  // needed to remove :invalid red border in Firefox
  box-shadow: none;

  &[type='number'],
  &[type='date'] {
    -moz-appearance: textfield;
    -webkit-appearance: none;
    &::-webkit-outer-spin-button,
    &::-webkit-inner-spin-button,
    &::-webkit-calendar-picker-indicator,
    &::-webkit-inner-spin-button,
    &::-webkit-clear-button {
      -webkit-appearance: none;
      appearance: none;
      margin: 0;
      display: none;
    }
  }
  &:focus {
    outline: 0;
  }
  &:not(:last-child) {
    padding-right: ${props => PADDING[props.size || 'regular']}px;
  }
  &::placeholder {
    color: ${wf(props => props.theme.colors.gray3)};
  }
  &:disabled::placeholder {
    color: ${wf(props => props.theme.colors.gray2)};
  }
`;

const Container = styled.div`
  ${textStyles('normal')}
  vertical-align: top;

  display: inline-flex;
  align-items: center;
  border: 1px solid
    ${wf(props =>
      props.isInvalid
        ? /* istanbul ignore next */ props.theme.colors.red
        : props.disabled
        ? props.theme.colors.gray1
        : props.theme.colors.gray3
    )};
  transition: border 0.2s ease-in-out;
  border-radius: 2px;
  padding: 0 ${props => PADDING[props.size || 'regular']}px;
  height: ${props => HEIGHT[props.size || 'regular']}px;
  background-color: transparent;
  color: ${wf(props =>
    props.disabled ? props.theme.colors.gray2 : props.theme.colors.primary
  )};

  &:focus-within {
    border-color: ${wf(props =>
      props.isInvalid
        ? /* istanbul ignore next */ props.theme.colors.red
        : props.theme.colors.selected
    )};
  }
`;

const InputText: React.StatelessFunctionalComponent<Props> = styled(
  React.forwardRef(
    ({ onChange, validationErrorMessage, ...props }: Props, ref) => {
      const input = React.createRef();
      return (
        <InvalidHandler errorMessage={validationErrorMessage}>
          {(getInputProps, isInvalid) => (
            <Container {...omit(props)(INPUT_ATTRIBUTES)} isInvalid={isInvalid}>
              <Input
                ref={mergeRefs(input, ref)}
                {...include(props)([...INPUT_ATTRIBUTES, 'disabled'])}
                {...getInputProps({ onChange })}
              />
              {props.renderAddon &&
                props.renderAddon({
                  onClick: () =>
                    input.current
                      ? input.current.focus()
                      : /* istanbul ignore next */ noop(),
                })}
            </Container>
          )}
        </InvalidHandler>
      );
    }
  )
)();

// @component
export default InputText;
