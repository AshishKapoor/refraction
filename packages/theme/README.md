This package provides a set of colors, sizes, and font styles, used to style
the applications created at Quid.

The default export of the package is an Emotion [theme provider][theme-provider]
that should wrap any application to make sure the other Quid components
are properly styled with the resources provided by this package.

Additionally, you can import the `colors`, `sizes`, `textStyles`, and `themes`
named exports to directly reference the variables.

### Usage

Wrap your component with `ThemeProvider` and, optionally, define the desired
theme (`light` or `dark`, `light` is default):

```jsx static
import ThemeProvider from '@quid/theme';

<ThemeProvider theme="light">
  <Button>I'll use the light theme</Button>
</ThemeProvider>;
```

Once a styled component is wrapped by the ThemeProvider, it can access
the current theme from the `props.theme` property.

Additionally, inside the theme object, you will find a `current` property
that will be set to `light` or `dark` depending by the active theme.  
This can be useful to conditionally style a component depending by its theme.

```jsx static
import styled from '@emotion/styled';
import ThemeProvider from '@quid/theme';

const Test = styled.span`
  color: ${props => props.theme.primary};
  background-color: ${props =>
    props.theme.current === 'dark' ? 'black' : 'white'};
`;

<ThemeProvider theme="light">
  <Test>Foobar</Test>
</ThemeProvider>;
```

### `withFallback`

The `withFallback` function is available to make it possible to create
your components and not have to worry of the case when no ThemeProvider
is defined to provide a theme to your component.

It will take care to set the light theme as default theme if no one is set.

```jsx static
import styled from '@emotion/styled';
import { withFallback as wf } from '@quid/theme';

const Button = styled.button`
  /* primary will always be defined no metter what */
  color: ${wf(props => props.theme.primary)};
`;
```

### `textStyles`

This utility will provide a set of CSS-in-JS compatible rules
to style a text according to the Quid UI guidelines.

The available styles are:

- `xlarge`, `large`, `normal` (text sizes)
- `title`, `body` (text sizes with serif font family)
- `bold`, `regular` (font weight)
- `secondary`, `disabled`, `link`, `highlighted` (text color)

```jsx static
import styled from '@emotion/styled';
import { textStyles } from '@quid/theme';

// This paragraph will have bold text, with the secondary text color
const Paragraph = styled.p`
  ${textStyles('bold', 'secondary')};
`;
```

### Font Faces

This package exposes two font-faces, IBM Plex (Sans and Serif), and the
iconic Quid Icons.

The first is used as font family of all our texts, while the second is
an iconic font family used to provide our icons.

You can import both of them in your project with:

```js static
import '@quid/theme/fonts/ibmplex';
import '@quid/theme/fonts/icons';
```

Once imported, they will be globally available for all the consumer components.

[theme-provider]: https://emotion.sh/docs/emotion-theming
