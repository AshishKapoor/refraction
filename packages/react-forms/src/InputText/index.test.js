/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import * as React from 'react';
import { mount } from 'enzyme';
import { colors } from '@quid/theme';
import InputText from '.';

it('renders the expected markup', () => {
  const wrapper = mount(<InputText data-foo="bar" />);
  expect(wrapper).toMatchSnapshot();
});

it('renders a disabled input', () => {
  const wrapper = mount(<InputText disabled />);
  expect(wrapper).toHaveStyleRule('color', colors.gray2);
});

it('renders an input with addon', () => {
  const Addon = ({ onClick }) => <div onClick={onClick}>foo</div>;
  const wrapper = mount(<InputText renderAddon={Addon} />);
  expect(wrapper.find('div').at(1)).toMatchSnapshot();
});

it('renders a small input', () => {
  const wrapper = mount(<InputText size="small" />);
  expect(wrapper).toHaveStyleRule('height', '24px');
});

it('renders a large input', () => {
  const wrapper = mount(<InputText size="large" />);
  expect(wrapper).toHaveStyleRule('height', '50px');
});

it('focuses the input on click on the input addon', () => {
  const handleFocus = jest.fn();
  const Addon = props => <div tabIndex={0} {...props} />;
  const wrapper = mount(
    <InputText renderAddon={({ onClick }) => <Addon onClick={onClick} />} />
  );
  wrapper
    .find('input')
    .getDOMNode()
    .addEventListener('focus', handleFocus);

  wrapper.find(Addon).simulate('click');
  expect(handleFocus).toHaveBeenCalled();
});
