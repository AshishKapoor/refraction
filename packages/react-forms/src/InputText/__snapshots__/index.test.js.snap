// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an input with addon 1`] = `
<div
  onClick={[Function]}
>
  foo
</div>
`;

exports[`renders the expected markup 1`] = `
.emotion-3 {
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 14px;
  line-height: 1.57;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 1px solid #C7CCD1;
  -webkit-transition: border 0.2s ease-in-out;
  transition: border 0.2s ease-in-out;
  border-radius: 2px;
  padding: 0 10px;
  height: 30px;
  background-color: #FFFFFF;
  color: #121212;
}

.emotion-3:focus-within {
  border-color: #00c1bb;
}

.emotion-0 {
  all: unset;
  min-width: 0;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  box-shadow: none;
}

.emotion-0[type='number'],
.emotion-0[type='date'] {
  -moz-appearance: textfield;
  -webkit-appearance: none;
}

.emotion-0[type='number']::-webkit-outer-spin-button,
.emotion-0[type='date']::-webkit-outer-spin-button,
.emotion-0[type='number']::-webkit-inner-spin-button,
.emotion-0[type='date']::-webkit-inner-spin-button,
.emotion-0[type='number']::-webkit-calendar-picker-indicator,
.emotion-0[type='date']::-webkit-calendar-picker-indicator,
.emotion-0[type='number']::-webkit-inner-spin-button,
.emotion-0[type='date']::-webkit-inner-spin-button,
.emotion-0[type='number']::-webkit-clear-button,
.emotion-0[type='date']::-webkit-clear-button {
  -webkit-appearance: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0;
  display: none;
}

.emotion-0:focus {
  outline: 0;
}

.emotion-0:not(:last-child) {
  padding-right: 10px;
}

<InputText
  data-foo="bar"
>
  <Container
    className="emotion-5 emotion-2"
    data-foo="bar"
    isInvalid={false}
  >
    <div
      className="emotion-2 emotion-3 emotion-4"
      data-foo="bar"
    >
      <Input
        onChange={[Function]}
        onInvalid={[Function]}
      >
        <input
          className="emotion-0 emotion-1"
          onChange={[Function]}
          onInvalid={[Function]}
        />
      </Input>
    </div>
  </Container>
</InputText>
`;
